File: Ashley/src/test/java/com/badlogic/ashley/systems/IntervalSystemTest.java
package com.badlogic.ashley.systems;

import static org.junit.Assert.assertEquals;

import com.badlogic.ashley.core.Engine;

import org.junit.Test;

public class IntervalSystemTest {
    private static final float deltaTime = 0.1f;

    @Test
    public void intervalSystem() {
        Engine engine = new Engine();
        IntervalSystemSpy intervalSystemSpy = new IntervalSystemSpy();

        engine.addSystem(intervalSystemSpy);

        for (int i = 1; i <= 10; ++i) {
            engine.update(deltaTime);
            assertEquals(i / 2, intervalSystemSpy.numUpdates);
        }
    }

    @Test
    public void testGetInterval() {
        IntervalSystemSpy intervalSystemSpy = new IntervalSystemSpy();
        assertEquals(deltaTime * 2.0f, intervalSystemSpy.getInterval(), 0);
    }

    private static class IntervalSystemSpy extends IntervalSystem {
        public int numUpdates;

        public IntervalSystemSpy() {
            super(deltaTime * 2.0f);
        }

        @Override
        protected void updateInterval() {
            ++numUpdates;
        }
    }
}

File: Ashley/src/test/java/com/badlogic/ashley/systems/IntervalSystemTest.kt
package com.badlogic.ashley.systems

import com.badlogic.ashley.core.Engine
import org.junit.Assert.assertEquals
import org.junit.Test

class IntervalSystemTest {

    @Test
    fun intervalSystem() {
        val engine = Engine()
        val intervalSystemSpy = IntervalSystemSpy()
        engine.addSystem(intervalSystemSpy)

        for (i in 1..10) {
            engine.update(DELTA_TIME)
            assertEquals(i / 2, intervalSystemSpy.numUpdates)
        }
    }

    @Test
    fun testGetInterval() {
        val intervalSystemSpy = IntervalSystemSpy()
        assertEquals(DELTA_TIME * 2.0F, intervalSystemSpy.getInterval(), 0F)
    }

    private class IntervalSystemSpy : IntervalSystem(DELTA_TIME * 2.0F) {
        var numUpdates = 0

        override fun updateInterval() {
            numUpdates++
        }
    }

    companion object {
        private const val DELTA_TIME = 0.1F
    }
}
